readme.md
# 轻量级 后台数据模拟（fe-mock-data)
  [github地址](https://github.com/RetinaLI/fe-mock-data)
## 特点
  + 轻量级，开发配置都很简单，提供初始化模版生成，只需要安装node和npm，不需要额外的比如mongodb、redis等等
  + 数据主要基于json文件，也可以自定义返回
  + 支持websocket，暂不支持restful接口
  + 灵活，提供所有配置项，也可以自定义
  + 模拟真实的数据请求流程，方便在开发过程中检查请求参数和请求头等信息是否正确

## 使用场景
  + 后端接口短时间不能开发完成，前端需要模拟数据，方便开发、调试
  + 在需要测试特定数据下的系统运行情况，且后端不方便提供模拟场景时，可以简单搭建一个数据环境

## 快速开始

   在开始之前，假设你已经安装了node(v8.x及以上)。

### 全局安装npm包
  	npm i fe-mock-data -g

### 使用命令行生成模版
  	fe-mock-data init

   * 会生成`mock-data`文件夹，里面包含一个`json`文件夹和一个`config.js`文件，仅供参考。
   * **注意：**`mock-data`文件夹、`json`文件夹、`config.js`均可以放置任意路径，只需要修改启动参数和配置路径即可。
   * 如果已经有mock-data文件夹，程序会自动比对，不会覆盖现有文件

### 启动

   	fe-mock-data run

   * 命令行提示 `mock server is ok!` 即为成功

### 完整启动

	fe-mock-data run --path  mock-data/config.js --port 4200  [--watch]

* `path`：路由配置文件，默认 mock-data/config.js
* `port`：端口，默认 4200
* `watch`：监听config修改并热启动，默认不启动

## 配置文件 config.js 的语法
   [进入](https://github.com/RetinaLI/fe-mock-data/blob/master/mock-data/config.js)

## 注意事项： 配置文件 config.js,在webpack + babel(vue/react/angular)项目中，其他模块引用方式
  1. 如果使用`export default {}`方式导出模块，引入方式;
  
    ```
    import xxx from '../mock-data/config.js';
    ```

  2. 如果使用`module.exports`方式导出模块，引用方式[可能会报错：export与module.exports不能混用];
  
    ```
      import xxx from '../mock-data/config.js';
    ```
    
    + 防止报错方法：比如修改`.babelrc`文件：

     ```
     plugins: [
      ...,
      "transform-runtime",
      "transform-es2015-modules-commonjs"
     ]
     ```
     两个配置项必须同时出现或者同时不出现。
  3. **特别注意：** 当使用websocket的renderFn进行自定义时，必须加上判断：ws.readyState === 1，如果有定时任务，也需要注意清除定时任务，比如：

    ```
      let intervalId = null;

      // websocket，自定义返回
      getWsDataByRenderFn: {
        url: '/ws/data/custom-render',
        method: 'ws',
        json: 'mock-data/json/data.json',
        renderFn: function(dataRes, ws, req, ext) {
          clearInterval(intervalId);
          intervalId = setInterval(() => {
            if (ws.readyState === 1) {
              ws.send(JSON.stringify(dataRes));
            } else {
              clearInterval(intervalId);
            }
          }, 4000);
        }
      }

    ```

## 更多命令
    fe-mock-data -h
    fe-mock-data -V


CHANGELOG.md
# feature

暂无

# 2.0.15
   支持routes中json配置项直接写json格式数据

# 2.0.14
   es6导入方式bug修复

# 2.0.12

1. **add**: 支持热启动：监听配置文件的修改并自动重新启动。启动命令参数：--watch，默认不启动

  ```
     fe-mock-data run --watch
  ```

  **注意：** 当使用`require('fe-mock-data').start({app: app})` 方式启动时，不会启动热启动。

# 2.0.11

1. 新增对es6导入方式import编译: 支持两种导入方式 export default {} 或者 module.exports方式;

# 2.0.10

1. fix mock-data/config.js 一个示例bug

# 2.0.9

1. **add:** 可以与webpack的webpack-dev-server插件集成，集成方式：

  ```
      const mock = require('fe-mock-data');

      ...

      devServer: {
            ...,

         // 由于webpack-dev-server 本身就是express服务器，可以直接使用
         before(app) {
               mock.start({
                  app: app,
                  port: 4200,
                  path: 'mock-data/config.js'
               });
         }
       }
  ```

2. **add:** 扩充引用**fe-mock-data**时的提供的方法

   ```
   const mock = require('fe-mock-data');

   // 分页获取数据方法
   mock.getPageData(arr, pageIndex, pageSize, startPageIndex)

   // 可以自定义启动服务，可以使用该方法和webpack-dev-server集成
   mock.start({
      app: app,  // express app
      port: 4200,
      path: 'mock-data/config.js'
   });

   //config文件中的某一项配置，会验证合法性、补足默认项
   mock.parseRouteConfig(routeConfig);

   // routeConfig：config文件中的某一项配置；
   // req 需要一个{ body?:{}, query?:{} }类似的结构，可以用来模拟传递的请求参数
   mock.getResponseData(routeConfig, req /* express request */);

   ```

3. **add:** 添加单元测试，进入**test**目录，执行**node test.js**

4. **update:** 命令行提示信息优化，提升健壮性

4. **update:** 规范了websocket自定义的写法，并添加到示例中

   ```
      let intervalId = null;
      // websocket，自定义返回，
      // 注意：一旦提供了renderFn，那么插件不会启动后台轮询，需要自行启动，比如setInterval
      getWsDataByRenderFn: {
         url: '/ws/data/custom-render',
         method: 'ws',
         json: 'mock-data/json/data.json',
         renderFn: function(dataRes, ws, req, ext) {
            clearInterval(intervalId);
            intervalId = setInterval(() => {
               if (ws.readyState === 1) {
                  ws.send(JSON.stringify(dataRes));
               } else {
                  clearInterval(intervalId);
               }
            }, 3000);
         }
      }
   ```

## BREAKING CHANGES

1. **update:** 配置调整：

   * **callbackStatus**调整为**statusCode**
   * **paginationQueryConfig.pageIndex**调整为**paginationQueryConfig.pageIndexKey**
   * **paginationQueryConfig.pageSize**调整为**paginationQueryConfig.pageSizeKey**
   * **paginationQueryConfig.listName**调整为**paginationQueryConfig.listKey**

2. **remove:** **ext**不再提供**getPageData**方法，使用如下方式替换：
   ```
      renderFn: (data, req, res, ext) {
        const MOCK_LIB = require('fe-mock-data');
        data.list = MOCK_LIB.getPageData(data.list, req.body.pageIndex, req.body.pageSize, 1);
        res.status(200).send(dataRes);
      }
   ```

test.js
let http = require('http');
let config = require('./config');
let server = require('./lib/server');
const MOCK_LIB = require('./lib/main');
let colors = require('colors');

let request = function (url, method) {
  return new Promise(resolve => {
    if (method !== 'get' && method !== 'post') return resolve(1);
    let options = {
      hostname: 'localhost',
      port: config.port,
      path: url,
      agent: false,
      method: method.toUpperCase()
    };
    let req = http.request(options, (res) => {
      res.setEncoding('utf8');
      let data = '';
      res.on('data', (str) => {
        data += str;
      });
      res.on('end', () => {
        resolve({ data: data, statusCode: res.statusCode });
      });
    });
    req.on('error', (e) => {
      resolve({ data: e.message, statusCode: req.statusCode });
    });
    req.end();
  });
};

let testApi = async function (key, routeConfig) {
  let result = await request(routeConfig.url, routeConfig.method);
  let statusCode = result.statusCode;
  result = result.data;
  let sourceData = MOCK_LIB.getResponseData(routeConfig, { query: {}, body: {} });
  if (sourceData != null && JSON.stringify(sourceData) !== result) {
    console.info((`test data fail:  ${routeConfig.url} + ${key}`).red);
    return false;
  }
  if (statusCode !== routeConfig.callbackStatus) {
    console.info((`test statusCode fail:  ${routeConfig.url} + ${key}`).red);
    return false;
  }
  console.info((`test success:  ${routeConfig.url} + ${key}`).green);
  return true;
};

server.start().then(() => {
  let mockConfig = require('./mock-data/config');
  let keys = Object.keys(mockConfig.routes);
  let completeCount = 0;
  keys.forEach(async (key, i) => {
    let routeConfig = MOCK_LIB.parseRouteConfig(mockConfig.routes[key]);
    await testApi(key, routeConfig);
    completeCount ++;
    if (completeCount === keys.length) {
      console.info('test finally! exit');
      // process.exit(0);
    }
  });
});

test/ws.js
// let net = require('net');

// let socket = new net.Socket();

// socket.connect(4200, 'localhost');
// socket.on('data',function(data){
//   console.log('from server:'+ data);
//   //得到服务端返回来的数据
// });


const WebSocket = require('ws');

const ws = new WebSocket('ws://localhost:4200/ws/data');

ws.on('open', function open() {
  ws.send('something');
});

ws.on('message', function incoming(data) {
  console.log(data);
});

test/test.js
let http = require('http');
let fs = require('fs');
let path = require('path');
const querystring = require('querystring');
let config = require('../config');
let server = require('../lib/server');
const MOCK_LIB = require('../lib/main');
let colors = require('colors');
let mock = require('../bin/mock');

let requestWS = function (url) {
  return new Promise(resolve => {
    const WebSocket = require('ws');

    const ws = new WebSocket('ws://localhost:' + config.port + url);

    ws.on('open', function incoming(data) {
      ws.send('hello');
      console.info(url.green, 'open success'.green);
    });
    ws.on('message', function incoming(data) {
      resolve((data));
    });
    ws.on('close', function (data) {
      console.info(url.cyan, ' closed'.cyan);
    });
    ws.on('error', function (data) {
      console.info(url.red, ' connect error success'.red, data.red);
    });
  });
};

let request = function (url, method, params) {
  return new Promise(resolve => {
    const postData = querystring.stringify(params);
    let options = {
      hostname: 'localhost',
      port: config.port,
      path: url,
      agent: false,
      method: method.toUpperCase(),
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Buffer.byteLength(postData)
      }
    };
    let req = http.request(options, (res) => {
      res.setEncoding('utf8');
      let data = '';
      res.on('data', (str) => {
        if (method === 'ws') console.info(str);
        data += str;
      });
      res.on('end', () => {
        resolve({ data: data, statusCode: res.statusCode });
      });
    });
    req.on('error', (e) => {
      resolve({ data: e.message, statusCode: req.statusCode });
    });
    req.write(postData);
    req.end();
  });
};

let testApi = async function (key, routeConfig) {
  let params = {};

  if (routeConfig.paginationQueryConfig) {
    params[routeConfig.paginationQueryConfig.pageIndexKey] = 1;
    params[routeConfig.paginationQueryConfig.pageSizeKey] = 2;
  } else {
    params.pageIndex = 1;
    params.pageSize = 2;
  }

  let result;
  if (routeConfig.method === 'ws') {
    result = await requestWS(routeConfig.url);
  } else {
    result = await request(routeConfig.url, routeConfig.method, params);
    let statusCode = result.statusCode;
    result = result.data;
    if (statusCode !== routeConfig.statusCode) {
      console.info((`test statusCode fail:  ${routeConfig.url} + ${key}`).red);
      return false;
    }
  }

  if (!routeConfig.paginationQueryConfig)
    routeConfig.paginationQueryConfig = {
      startIndex: 1,
      listKey: 'data.list'
    };
  routeConfig = MOCK_LIB.parseRouteConfig(routeConfig);
  let sourceData = MOCK_LIB.getResponseData(routeConfig, { query: params, body: params });
  if (sourceData != null && JSON.stringify(sourceData) !== result) {
    console.info((`test data fail:  ${routeConfig.url} + ${key}`).red);

    console.info(result, sourceData);
    return false;
  }
  console.info((`test success:  ${routeConfig.url} + ${key}`).green);
  return true;
};

let testInitCommand = function () {
  let removefs = (_path) => {
    try {
      let stats = fs.statSync(_path);
      if (stats.isDirectory()) {
        let objs = fs.readdirSync(_path);
        objs.forEach(obj => {
          removefs(path.resolve(_path, obj));
        });
        fs.rmdirSync(_path);
      } else {
        fs.unlinkSync(_path);
      }
    } catch (ex) {
      throw ex;
    }
  };
  try {
    removefs(path.resolve(__dirname, 'mock-data'));
  } catch (ex) {
    console.info('test init command removefs fail'.red);
    return;
  }
  try {
    mock.init();
  } catch (ex) {
    console.info('test init command fail'.red);
    throw ex;
  }

  console.info('test init command success.'.green);
};

let testServerRun = function () {
  server.start().then(() => {
    let mockConfig = require('./mock-data/config');
    let keys = Object.keys(mockConfig.routes);
    let completeCount = 0;
    keys.forEach(async (key, i) => {
      let routeConfig = MOCK_LIB.parseRouteConfig(mockConfig.routes[key]);
      await testApi(key, routeConfig);
      completeCount++;
      if (completeCount === keys.length) {
        console.info('test finally! exit');
        process.exit(0);
      }
    });
  });
};

testInitCommand();
testServerRun();


index.js
#!/usr/bin/env node
if (process.argv.length > 1) {
  let config = require('./config');
  let program = require('commander');
  let packageInfo = require('./package.json');

  program
  .version(packageInfo.version)
  .option('-p, --path [path]', '配置文件路径', config.config)
  .option('-P, --port [port]', '配置端口号', config.port)
  .option('-W, --watch', '监听config修改并热启动，默认不启动');

  program
    .command('init')
    .description('初始化mock-server文件夹')
    .action(function() {
      let mock = require('./bin/mock.js');
      mock.init();
    });

  program
    .command('run')
    .description('运行模拟后台数据')
    .action(function() {
      let server = require('./lib/server');
      server.start({
        port: program.port || config.port,
        path: program.path || config.config,
        enableWatch: program.watch,
        app: null
      });
    });

  program.parse(process.argv);
}
module.exports = require('./lib/main');


lib/main.js
// 获取分页数据
function getPageData (json = null, pageIndex = 0, pageSize = 20, keyName = 'data', startPage = 0) {
  let result = getDataFromJsonFile(json);
  if (result) {
    let resJson = result;
    let list = resJson;
    let arr = keyName.split('.');
    let temp;

    arr.forEach((k, i) => {
      list = list[k];

      if (i === arr.length - 2) {
        temp = list;
      }
    });

    temp[arr[arr.length - 1]] = model.getPageData(list, pageIndex, pageSize, startPage);
    return resJson;
  }
}

let getRequestParams = function (req, method) {
  return method === 'get' ? req.query : req.body;
};

// 获取json全部数据，不分页
function getDataFromJsonFile (json) {
  let fs = require('fs');
  let path = require('path');
  if (typeof json === 'string') {
    return JSON.parse(fs.readFileSync(path.join(path.resolve('.'), json)));
  } else if ( typeof json === 'object') {
    return json
  }

}

// 返回数据公共代码提取
function getCommResult (routeConfig, req) {
  let config = routeConfig.paginationQueryConfig;
  let pIndex, pSize, pList, startIndex, dataRes;

  pIndex = config.pageIndexKey;
  pSize = config.pageSizeKey;
  pList = config.listKey;
  startIndex = config.startIndex;
  let nPageIndex = 0;
  let nPageSize = 20;
  let params = getRequestParams(req, routeConfig.method);

  nPageIndex = parseInt(params[pIndex]);
  nPageSize = parseInt(params[pSize]);

  if (!isNaN(nPageIndex) && !isNaN(nPageSize)) {
    dataRes = getPageData(routeConfig.json, nPageIndex, nPageSize, pList, startIndex);
  } else {
    dataRes = getDataFromJsonFile(routeConfig.json);
  }
  return dataRes;
}
// 返回dataRes
function getResponseData (routeConfig, req) {
  if (!routeConfig.json) return null;
  if (!routeConfig.paginationQueryConfig) {
    return getDataFromJsonFile(routeConfig.json);
  }
  return getCommResult(routeConfig, req);
}
const model = {
  getResponseData,
  parseRouteConfig (routeConfig) {
    routeConfig.method = ((routeConfig.method || 'get') + '').toLowerCase();
    if (routeConfig.method !== 'get' && routeConfig.method !== 'post' && routeConfig.method !== 'ws') { throw Error('配置请求方式错误，必须为get/post/ws其中一种') };

    routeConfig.url = routeConfig.url || '/';

    routeConfig.statusCode = routeConfig.statusCode || 200;
    if (isNaN(parseInt(routeConfig.statusCode))) { throw Error('状态码必须为数字') };

    if (routeConfig.method === 'ws') {
      routeConfig.interval = routeConfig.interval || 5000;
      if (!routeConfig.renderFn && isNaN(parseInt(routeConfig.interval))) { throw Error('轮训时间必须为数字，单位ms') };
    }
    if (routeConfig.paginationQueryConfig) {
      routeConfig.paginationQueryConfig.pageIndexKey = routeConfig.paginationQueryConfig.pageIndexKey || 'pageIndex';
      routeConfig.paginationQueryConfig.pageSizeKey = routeConfig.paginationQueryConfig.pageSizeKey || 'pageSize';
      routeConfig.paginationQueryConfig.startIndex = routeConfig.paginationQueryConfig.startIndex == null ? 0 :  routeConfig.paginationQueryConfig.startIndex;
      routeConfig.paginationQueryConfig.listKey = routeConfig.paginationQueryConfig.listKey || 'list';
    }
    // console.info(routeConfig.paginationQueryConfig.pageIndexKey);
    return routeConfig;
  },
  getPageData: function (arrayData, pageIndex, pageSize, startPage) {
    pageIndex = +pageIndex;
    pageSize = +pageSize;
    let pageData = [];
    if (arrayData && arrayData.length > 0) {
      if (startPage === 0) {
        pageData = arrayData.slice(pageSize * pageIndex, (pageIndex + 1) * pageSize);
      } else {
        pageData = arrayData.slice(pageSize * (pageIndex - 1), pageIndex * pageSize);
      }
      return pageData;
    }
  },
  start (opt) {
    if (!opt.app) throw 'need an app instance of express'
    let server = require('./server');
    server.start(opt);
  }
};

module.exports = model;

lib/server.js
#!/usr/bin/env node
let path = require('path');
let fs = require('fs');
let config = require('../config');
let colors = require('colors');
const MOCK_LIB = require('../lib/main');
let watch = require('node-watch');
let key = Date.now();
let babelEs2015 = require('babel-plugin-transform-es2015-modules-commonjs');

let watcher, app, server;

let getExtLib = function () {
  return {
    path: path,
    fs: fs,
    __dirname: __dirname
  }
};

let restartApp = function (opt) {
  console.info('mock server restarting!'.green);
  if (server) {
    console.info('old mock server closing!'.green);
    server.close(() => {
      console.info('old mock server closed!'.green);
      console.info('new mock server starting!'.green);
      start(opt);
    });
  } else {
    start(opt);
  }
};

let startWatchConfig = function (opt) {
  watcher = watch(path.dirname(path.join(path.resolve('.'), opt.path)), { filter: /\.js$/ }, function (evt, name) {
    console.info('file changed: '.blue + name.green);
    key = Date.now();
    delete require.cache[require.resolve(path.join(path.resolve('.'), opt.path))];
    restartApp(opt);
  });
  console.info('file watching: '.blue + path.dirname(path.join(path.resolve('.'), opt.path)).green);
};

let parseRoutes = function (serverConfig) {
  let routesMap = serverConfig.routes;
  console.info(('parse routes: (' + Object.keys(routesMap).length + '个)').bgGreen);
  Object.keys(routesMap).forEach(key => {
    let routeConfig = MOCK_LIB.parseRouteConfig(routesMap[key]);
    // 以websocket为分界
    if (routeConfig.method === 'ws') {
      app.ws(routeConfig.url, (ws, req) => {
        let dataRes = MOCK_LIB.getResponseData(routeConfig);
        if (routeConfig.renderFn) {
          try {
            routeConfig.renderFn(dataRes, ws, req, getExtLib());
          } catch (ex) {
            console.info(`error:`.bgRed + ` ${key} ${routeConfig.url} ${ex.message}`.red);
          }
        } else if (routeConfig.json) {
          dataRes = JSON.stringify(dataRes);
          let intervalId = setInterval(() => {
            try {
              if (ws.readyState === 1) {
                ws.send(dataRes);
              } else {
                clearInterval(intervalId);
              }
            } catch (ex) {
              console.info(ex.message.error);
              clearInterval(intervalId);
            }
          }, routeConfig.interval);

          ws.on('message', (msg) => {
            console.log("收到" + key + "发来的数据：" + msg);
          });

          ws.on('close', function () {
            clearInterval(intervalId);
          });
        };
      });
    } else {
      // 非websocket
      app[routeConfig.method](routeConfig.url, (req, res) => {
        let dataRes = MOCK_LIB.getResponseData(routeConfig, req);
        if (routeConfig.renderFn) {
          try {
            routeConfig.renderFn(dataRes, req, res, getExtLib());
          } catch (ex) {
            console.info(`error:`.bgRed + ` ${key} ${routeConfig.url} ${ex.message}`.red);
          }
        } else {
          res.status(routeConfig.statusCode).send(dataRes);
        }
      });
    }

    console.info(('     (' + routeConfig.method + ') ' + routeConfig.url).green);
  });
};

let initStaticPath = function (serverConfig) {
  if (!serverConfig.publicPath || !serverConfig.publicName) return;
  let express = require('express');
  app.use(serverConfig.publicName, express.static(serverConfig.publicPath));
};

let initExpressApp = function (opt) {
  if (!opt.app) {
    let express = require('express');
    app = express();
    app.use((req, res, next) => {
      res.header('Access-Control-Allow-Origin', '*');
      res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
      res.header('Access-Control-Allow-Methods', 'PUT,POST,GET,DELETE,OPTIONS');
      next();
    });
  } else {
    app = opt.app;
  }
  let bodyParser = require('body-parser');
  let expressWs = require('express-ws')(app);

  app = expressWs.app;
  app.use(bodyParser.urlencoded({
    extended: false,
    limit: '50mb'
  }));
};


let start = function (opt = {
  port: config.port,
  path: config.config,
  enableWatch: false,
  app: null
}) {
  let currentKey = key;
  opt.enableWatch = opt.app ? false : opt.enableWatch;
  // 进行import编译
  require('babel-register')({
    plugins: [babelEs2015],
  })
  let serverConfig = require(path.join(path.resolve('.'), opt.path));
  // 只支持 export default {} 或者 module.exports方式;
  serverConfig = serverConfig.default || serverConfig;
  initExpressApp(opt);
  initStaticPath(serverConfig);
  parseRoutes(serverConfig);

  return (new Promise((resolve) => {
    if (!opt.app) {
      if (key !== currentKey) return resolve(2);
      server = app.listen(opt.port, () => {
        resolve();
      });

      if (opt.enableWatch && !watcher) {
        startWatchConfig(opt);
      }
    } else {
      resolve();
    }
  })).then((result) => {
    if (result === 2) return;
    console.info(`mock server is ok!`.bgGreen, `http://localhost:${opt.port}`.underline.cyan);
    return true;
  }, () => {
    console.info(`mock server has failed!`.red);
    return false;
  });
};


module.exports = {
  start
};

mock-data/config.js
let intervalId = null;

export default {
  // publicPath + publicName 配置静态目录，参考express.static
  // app.use(publicName, express.static(publicPath));
  publicPath: 'mock-data/public',
  publicName: 'static',

  // 路由配置
  routes: {
    // 获取单个数据
    getSingleData: {
      url: '/data', // 默认 /
      method: 'get', // 默认 get
      statusCode: 200, // 默认 200
      json: {
        "data": {
          "name": "jakson"
        }
      }
    },
    // 返回404状态码
    getSingleDataWith404: {
      url: '/data/404',
      statusCode: 404,
      json: 'mock-data/json/data.json'
    },
    // 获取分页数据
    getList: {
      url: '/data/list',
      method: 'post',
      json: 'mock-data/json/list.json',
      paginationQueryConfig: {
        pageIndexKey: 'index', // 默认 pageIndex
        pageSizeKey: 'size', // 默认 pageSize
        startIndex: 1,  // 默认 0
        listKey: 'data.list' // 根据该配置路径查找具体的列表数据数组， 默认 list
      }
    },
    // 通过renderFn，自定义返回结果
    getListByRenderFn: {
      url: '/data/list-by-render',
      method: 'post',
      json: 'mock-data/json/list.json',
      renderFn: function (dataRes, req, res, ext) {
        /**
         * MOCK_LIB.getPageData: function (list: any[], pageIndex: number, pageSize: number, startIndex: number) return list: any[];
         */
        const MOCK_LIB = require('fe-mock-data');
        dataRes.data.list = MOCK_LIB.getPageData(dataRes.data.list, req.body.pageIndex, req.body.pageSize, 1);
        res.status(200).send(dataRes);
      }
    },
    /**
     * 通过renderFn + ext，自定义返回结果
     * ext: {   // 开放一些nodejs api和环境变量
     *  path: require('path'),
     *  fs: require('fs'),
     *  __dirname: __dirname
     * }
     *  */
    getDataWithoutJson: {
      url: '/data/no-json',
      method: 'post',
      renderFn: function (dataRes, req, res, ext) {
        let content = ext.fs.readFileSync(ext.path.join(ext.path.resolve(''), 'mock-data/json/data.json'));
        res.status(200).send(JSON.parse(content));
      }
    },
    // 获取一个文件
    getFile: {
      url: '/file',
      method: 'get',
      renderFn: function (dataRes, req, res, ext) {
        res.status(200).sendFile(ext.path.join(ext.path.resolve(''), 'mock-data/json/identify.png'));
      }
    },
    // websocket， 会重复返回json里面的数据
    getWsData: {
      url: '/ws/data',
      method: 'ws',
      interval: 3000,  // 默认5000ms
      json: 'mock-data/json/data.json'
    },
    // websocket，自定义返回，
    // 注意：一旦提供了renderFn，那么插件不会启动后台轮询，需要自行启动，比如setInterval
    getWsDataByRenderFn: {
      url: '/ws/data/custom-render',
      method: 'ws',
      json: 'mock-data/json/data.json',
      renderFn: function(dataRes, ws, req, ext) {
        clearInterval(intervalId);
        intervalId = setInterval(() => {
          if (ws.readyState === 1) {
            ws.send(JSON.stringify(dataRes));
          } else {
            clearInterval(intervalId);
          }
        }, 3000);
      }
    }
  }
};